From c119920b7d0e311974be5f311b3b52bf0fe48518 Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Wed, 10 Dec 2014 14:30:28 +0700
Subject: [PATCH 1/3] Fix Quick Settings for MSim [2/2]

- Use active data sim for setDataEnabled/getDataEnabled
- Also update system setting  for setDataEnabled so that the correct
  state is displayed in Settings

Change-Id: I1c72d9955b1d068ef276386b761f7aa2d664fdc2
---
 src/com/android/phone/PhoneInterfaceManager.java | 23 +++++++++++++++--------
 1 file changed, 15 insertions(+), 8 deletions(-)

diff --git a/src/com/android/phone/PhoneInterfaceManager.java b/src/com/android/phone/PhoneInterfaceManager.java
index 6c23a8e..3ec936b 100644
--- a/src/com/android/phone/PhoneInterfaceManager.java
+++ b/src/com/android/phone/PhoneInterfaceManager.java
@@ -1146,17 +1146,13 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
 
     // FIXME: subId version needed
     public boolean enableDataConnectivity() {
-        enforceModifyPermission();
-        long subId = SubscriptionManager.getDefaultDataSubId();
-        getPhone(subId).setDataEnabled(true);
+        setDataEnabled(true);
         return true;
     }
 
     // FIXME: subId version needed
     public boolean disableDataConnectivity() {
-        enforceModifyPermission();
-        long subId = SubscriptionManager.getDefaultDataSubId();
-        getPhone(subId).setDataEnabled(false);
+        setDataEnabled(false);
         return true;
     }
 
@@ -1949,7 +1945,17 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
     @Override
     public void setDataEnabled(boolean enable) {
         enforceModifyPermission();
-        mPhone.setDataEnabled(enable);
+        long subId = SubscriptionManager.getDefaultDataSubId();
+        int phoneId = SubscriptionManager.getPhoneId(subId);
+
+        if (TelephonyManager.getDefault().isMultiSimEnabled()) {
+            // If the non-subid API is used, we want to update settings as well
+            // (see DataUsageSummary.java in packages/apps/Settings)
+            android.provider.Settings.Global.putInt(mPhone.getContext().getContentResolver(),
+                    android.provider.Settings.Global.MOBILE_DATA + phoneId, enable ? 1 : 0);
+        }
+
+        getPhone(phoneId).setDataEnabled(enable);
     }
 
     /**
@@ -1982,7 +1988,8 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
             mApp.enforceCallingOrSelfPermission(android.Manifest.permission.MODIFY_PHONE_STATE,
                     null);
         }
-        return mPhone.getDataEnabled();
+        long subId = SubscriptionManager.getDefaultDataSubId();
+        return getPhone(subId).getDataEnabled();
     }
 
     @Override
-- 
1.9.3 (Apple Git-50)


From 96784b3a2f9f07dce4cd5802a93d31446a66710e Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Sun, 23 Nov 2014 15:48:23 +0700
Subject: [PATCH 2/3] Telephony: GSM: remove 'Carrier Settings' regardless of
 csp_enabled

Carrier Settings should be removed when config_carrier_settings_enable
is false. csp_enabled has nothing to do with it.

Fixes "Carrier Settings" appearing and crashing when the user clicks
on it.

Change-Id: I31c508f2353eb15a096aa279e71c4be4ce89abb8
---
 src/com/android/phone/GsmUmtsOptions.java | 17 ++++++++---------
 1 file changed, 8 insertions(+), 9 deletions(-)

diff --git a/src/com/android/phone/GsmUmtsOptions.java b/src/com/android/phone/GsmUmtsOptions.java
index 463156d..25e2228 100755
--- a/src/com/android/phone/GsmUmtsOptions.java
+++ b/src/com/android/phone/GsmUmtsOptions.java
@@ -143,15 +143,14 @@ public class GsmUmtsOptions {
                 log("[CSP] Disabling Operator Selection menu.");
                 mPrefScreen.removePreference(mButtonOperatorSelectionExpand);
             }
-
-            // Read platform settings for carrier settings
-            final boolean isCarrierSettingsEnabled = mPrefActivity.getResources().getBoolean(
-                    R.bool.config_carrier_settings_enable);
-            if (!isCarrierSettingsEnabled) {
-                Preference pref = mPrefScreen.findPreference(BUTTON_CARRIER_SETTINGS_KEY);
-                if (pref != null) {
-                    mPrefScreen.removePreference(pref);
-                }
+        }
+        // Read platform settings for carrier settings
+        final boolean isCarrierSettingsEnabled = mPrefActivity.getResources().getBoolean(
+                R.bool.config_carrier_settings_enable);
+        if (!isCarrierSettingsEnabled) {
+            Preference pref = mPrefScreen.findPreference(BUTTON_CARRIER_SETTINGS_KEY);
+            if (pref != null) {
+                mPrefScreen.removePreference(pref);
             }
         }
         if (!mRemovedAPNExpand) {
-- 
1.9.3 (Apple Git-50)


From 9d1fd9572c1f07c9788249a36e7c517694f6de04 Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Sun, 23 Nov 2014 15:57:20 +0700
Subject: [PATCH 3/3] Telephony: MSim: enable preferred mode button for all
 phones

HACK to allow changing network mode on multi-SIM

Change-Id: I79cd6dfb1653c9c1ef14b8965c9af2bd0f5792cb
---
 res/values/strings.xml                             | 10 ++++++++
 res/xml/msim_network_sub_setting.xml               |  4 ++--
 .../phone/msim/MSimMobileNetworkSubSettings.java   | 27 +++++++---------------
 3 files changed, 20 insertions(+), 21 deletions(-)

diff --git a/res/values/strings.xml b/res/values/strings.xml
index 40a0aca..27667c4 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -551,6 +551,16 @@
     <string name="preferred_network_mode_gsm_only_choice">GSM only</string>
     <string name="preferred_network_mode_gsm_wcdma_preferred_choice">GSM/WCDMA preferred</string>
 
+    <string-array name="preferred_network_mode_choices_2" translatable="false">
+        <item>@string/preferred_network_mode_wcdma_only_choice</item>
+        <item>@string/preferred_network_mode_gsm_only_choice</item>
+        <item>@string/preferred_network_mode_gsm_wcdma_preferred_choice</item>
+    </string-array>
+    <string-array name="preferred_network_mode_values_2"  translatable="false">
+        <item>"2"</item>
+        <item>"1"</item>
+        <item>"0"</item>
+    </string-array>
 
     <string-array name="preferred_network_mode_choices" translatable="false">
         <item>@string/preferred_network_mode_td_scdma_lte_gsm_wcdma_cdma_evdo_choice </item>
diff --git a/res/xml/msim_network_sub_setting.xml b/res/xml/msim_network_sub_setting.xml
index 6cd0eb5..72101de 100644
--- a/res/xml/msim_network_sub_setting.xml
+++ b/res/xml/msim_network_sub_setting.xml
@@ -40,8 +40,8 @@
         android:key="preferred_network_mode_key"
         android:title="@string/preferred_network_mode_title"
         android:summary="@string/preferred_network_mode_summary"
-        android:entries="@array/preferred_network_mode_choices"
-        android:entryValues="@array/preferred_network_mode_values"
+        android:entries="@array/preferred_network_mode_choices_2"
+        android:entryValues="@array/preferred_network_mode_values_2"
         android:dialogTitle="@string/preferred_network_mode_dialogtitle" />
 
 </PreferenceScreen>
diff --git a/src/com/android/phone/msim/MSimMobileNetworkSubSettings.java b/src/com/android/phone/msim/MSimMobileNetworkSubSettings.java
index 3a5a23c..a657981 100644
--- a/src/com/android/phone/msim/MSimMobileNetworkSubSettings.java
+++ b/src/com/android/phone/msim/MSimMobileNetworkSubSettings.java
@@ -212,32 +212,21 @@ public class MSimMobileNetworkSubSettings extends PreferenceActivity
             mUPLMNPref.getIntent().putExtra(PhoneConstants.SUBSCRIPTION_KEY, mPhone.getPhoneId());
         }
 
-        boolean isLteOnCdma = mPhone.getLteOnCdmaMode() == PhoneConstants.LTE_ON_CDMA_TRUE;
-        if (getResources().getBoolean(R.bool.world_phone) == true) {
-            // set the listener for the mButtonPreferredNetworkMode list preference so we can issue
-            // change Preferred Network Mode.
-            mButtonPreferredNetworkMode.setOnPreferenceChangeListener(this);
+        // set the listener for the mButtonPreferredNetworkMode list preference so we can issue
+        // change Preferred Network Mode.
+        mButtonPreferredNetworkMode.setOnPreferenceChangeListener(this);
 
-            //Get the networkMode from Settings.System and displays it
-            int settingsNetworkMode = getPreferredNetworkMode();
-            mButtonPreferredNetworkMode.setValue(Integer.toString(settingsNetworkMode));
+        //Get the networkMode from Settings.System and displays it
+        int settingsNetworkMode = getPreferredNetworkMode();
+        mButtonPreferredNetworkMode.setValue(Integer.toString(settingsNetworkMode));
+
+        if (getResources().getBoolean(R.bool.world_phone) == true) {
             mCdmaOptions = new CdmaOptions(this, prefSet, mPhone);
             mGsmUmtsOptions = new GsmUmtsOptions(this, prefSet, mPhone.getPhoneId());
         } else {
-            if (!isLteOnCdma) {
-                prefSet.removePreference(mButtonPreferredNetworkMode);
-            }
             int phoneType = mPhone.getPhoneType();
             if (phoneType == PhoneConstants.PHONE_TYPE_CDMA) {
                 mCdmaOptions = new CdmaOptions(this, prefSet, mPhone);
-                if (isLteOnCdma) {
-                    mButtonPreferredNetworkMode.setOnPreferenceChangeListener(this);
-
-                    int settingsNetworkMode = getPreferredNetworkMode();
-                    mButtonPreferredNetworkMode.setValue(
-                            Integer.toString(settingsNetworkMode));
-                }
-
             } else if (phoneType == PhoneConstants.PHONE_TYPE_GSM) {
                 mGsmUmtsOptions = new GsmUmtsOptions(this, prefSet, mPhone.getPhoneId());
             } else {
-- 
1.9.3 (Apple Git-50)

